src/controller/artifact/processor/default.go

```diff
@@ -16,44 +16,178 @@ package processor
 }
 
-func (d *defaultProcessor) GetArtifactType() string {
+func (d *defaultProcessor) GetArtifactType(ctx context.Context, artifact *artifact.Artifact) string {
 	// try to parse the type from the media type
-	strs := artifactTypeRegExp.FindStringSubmatch(d.mediaType)
+	strs := artifactTypeRegExp.FindStringSubmatch(artifact.MediaType)
 	if len(strs) == 2 {
 		return strings.ToUpper(strs[1])
 	}
 	// can not get the artifact type from the media type, return unknown
 	return ArtifactTypeUnknown
 }
-func (d *defaultProcessor) ListAdditionTypes() []string {
-	return nil
+func (d *defaultProcessor) ListAdditionTypes(ctx context.Context, artifact *artifact.Artifact) []string {
+	artifactConfig := artifact.ExtraAttrs
+	schemaMap, ok := artifactConfig[schemaV1.XHarborAttributes]
+	if !ok {
+		return nil
+	}
+	schema := &schemaV1.Schema{}
+	err := mapstructure.Decode(schemaMap, schema)
+	if err != nil {
+		log.Warningf("unsupported artifact config schema: %v", err)
+		return nil
+	}
+	additionTypes := make([]string, 0)
+	for _, addition := range schema.Additions {
+		additionTypes = append(additionTypes, addition.Name)
+	}
+	return additionTypes
 }
-func (d *defaultProcessor) AbstractMetadata(ctx context.Context, manifest []byte, artifact *artifact.Artifact) error {
-	// do nothing currently
-	// we can extend this function to abstract the metadata in the future if needed
+func (d *defaultProcessor) AbstractMetadata(ctx context.Context, artifact *artifact.Artifact, manifest []byte) error {
+	if artifact.ManifestMediaType != v1.MediaTypeImageManifest && artifact.ManifestMediaType != schema2.MediaTypeManifest {
+		return nil
+	}
+	// get manifest
+	mani := &v1.Manifest{}
+	if err := json.Unmarshal(manifest, mani); err != nil {
+		log.Warningf("unsupported manifest: %v", err)
+		return nil
+	}
+	// get config layer
+	_, blob, err := d.RegCli.PullBlob(artifact.RepositoryName, mani.Config.Digest.String())
+	if err != nil {
+		log.Warningf("get config layer error: %v", err)
+		return nil
+	}
+	defer blob.Close()
+	// parse metadata from config layer
+	metadata := map[string]interface{}{}
+	if err := json.NewDecoder(blob).Decode(&metadata); err != nil {
+		log.Warningf("parse config layer error: %v", err)
+		return nil
+	}
+
+	// populate all metadata into the ExtraAttrs first
+	// if there are xHarborAttribute in artifact config
+	// metadata will be processed in specific rules
+	artifact.ExtraAttrs = metadata
+
+	// parse schema
+	schemaMap, ok := metadata[schemaV1.XHarborAttributes]
+	if !ok {
+		return nil
+	}
+	schema := &schemaV1.Schema{}
+	err = mapstructure.Decode(schemaMap, schema)
+	if err != nil {
+		log.Warningf("unsupported artifact config schema: %v", err)
+		return nil
+	}
+	// check schema version
+	if schema.SchemaVersion != schemaV1.SchemaVersionV1 {
+		log.Warningf("unsupported artifact config schema version %d", schema.SchemaVersion)
+		return nil
+	}
+	// check icon url
+	_, err = url.ParseRequestURI(schema.Icon)
+	if err != nil {
```

Comment:
The logic here is a bit weird.
Why do not copy the original schema text into a var before and use defer func to override the `metadata[schemaV1.XHarborAttributes] `?

Reply:
```metadata[schemaV1.XHarborAttributes]``` is defined as ```map[string]interface{}``` Because ```artifact.ExtraAttrs``` is defined as ```map[string]interface{}```. If icon is invalid, I need to set it as a empty string. And convert schema struct to ```map[string]interface{}```. What I do here is to covert data to ```map[string]interface{}```.

Reply:
I think my question is why the invalid icon URL influences the whole data processing? We just need to ignore it (set it to empty str) and continue the processing.

Reply:
What I do here is to set icon to empty str in metadata. Set icon empty in schema struct and convert it to metadata.If just continue to process, icon url still exist in original metadata.

Reply:
Just ingore it in schema struct is useless, we also need to ignore it in metadata, which will be used in extra_attr.

Reply:
Discussed with @steven-zou, we will not processor icon URL here. Maybe we will start another PR to process icon